### PR TITLE
[Impeller] Compute geometry for drawPoints in vertex shader, with fallback for optimized CPU routine.

### PR DESCRIPTION
**Pull request by jonahwilliams**
Example of using disabling rasterization and using a vertex shader for computing circle geometry. This is only wired up in Metal, but could be supported in GLES 3.0 / WebGL with  [Transform_Feedback](https://www.khronos.org/opengl/wiki/Transform_Feedback)

<google-sheets-html-origin><style type="text/css"></style>

IMpeller Before TEss Fix |   | 500ms
-- | -- | --
Impeller TOT |   | 52ms
Impeller CPU |   | 14ms
Impeller Vert |   | 2.4ms

Benchmark is pending here: https://github.com/flutter/flutter/pull/126728

### CODE REVIEW

File: impeller/entity/geometry.cc [hunk 1]
```diff
@@ -883,4 +892,135 @@ std::optional<Rect> RRectGeometry::GetCoverage(const Matrix& transform) const {
...
+  if (scaled_radius < 4.0) {
+    return 8;
+  }
+  if (scaled_radius < 16) {
+    return 16;
+  }
+  if (scaled_radius < 32) {
+    return 32;
+  }
+  return 64;
```
Comment (jonahwilliams) [inline impeller/entity/geometry.cc:1012]: TODO

File: impeller/renderer/pipeline_builder.h [hunk 2]
```diff
@@ -16,6 +16,15 @@
 
 namespace impeller {
 
+class NonRenderingFragment {
+ public:
+  static constexpr std::string_view kLabel = "NonRendering";
+  static constexpr std::string_view kEntrypointName = "non_rendering_main";
+  static constexpr ShaderStage kShaderStage = ShaderStage::kFragment;
+  static constexpr std::string_view kGeneratorName = "NonRendering";
```
Comment (jonahwilliams) [inline impeller/renderer/pipeline_builder.h:24]: TODO: right now the debug info comes from the fragment name but for non-rasterizing pipelines it should come from the vertex name.

File: impeller/renderer/pipeline_builder.h [hunk 3]
```diff
@@ -52,35 +61,44 @@ struct PipelineBuilder {
...
-                       << "'.";
-        return false;
+
+      std::shared_ptr<const ShaderFunction> fragment_function;
+      if (!std::is_same<FragmentShader, NonRenderingFragment>::value) {
+        fragment_function = context.GetShaderLibrary()->GetFunction(
+            FragmentShader::kEntrypointName, ShaderStage::kFragment);
+      }
+
+      if (!std::is_same<FragmentShader, NonRenderingFragment>::value) {
```
Comment (jonahwilliams) [inline impeller/renderer/pipeline_builder.h:87]: TODO: make sure to validate vertex shader still

File: impeller/entity/geometry.h [hunk 4]
```diff
@@ -290,4 +294,38 @@ class RRectGeometry : public Geometry {
   FML_DISALLOW_COPY_AND_ASSIGN(RRectGeometry);
 };
 
+class PointFieldGeometry : public Geometry {
```
Comment (jonahwilliams) [inline impeller/entity/geometry.h:297]: TODO: figure out where to put the fallback to CPU computation. Could just add a fallback here?

File: impeller/entity/geometry.cc (hunk 1 above)
Comment (dnfield) [reply impeller/entity/geometry.cc:1012]: See `SkNextPow2` and `SkNextLog2`.

File: impeller/entity/geometry.h (hunk 4 above)
Comment (jonahwilliams) [reply impeller/entity/geometry.h:267]: Done

**Comment by jonahwilliams** [comment]
This is working and has the correct fallbacks, last remaining task is to fix the subdivision curve.

File: impeller/entity/geometry.cc (hunk 1 above)
Comment (dnfield) [reply impeller/entity/geometry.cc]: (Oh, I see, this is supposed to be modelling a curve not just always getting the next power of 2)

**Comment by jonahwilliams** [comment]
Closing in favor of compute based approach

**Pull request closed without merging**